oxidation/libparsec/crates/core_fs/src/file_operations.rs

```diff
@@ -0,0 +1,740 @@
+/// - a destination block, represented as a single chunk
+/// - a write back boolean indicating the new chunk must be written
+/// - a `HashSet` of chunk IDs that must cleaned up from the storage
+pub fn prepare_reshape(
+    manifest: &LocalFileManifest,
+) -> impl Iterator<Item = (u64, Vec<Chunk>, Chunk, bool, HashSet<ChunkID>)> + '_ {
+    // Loop over blocks
+    manifest
+        .blocks
+        .iter()
+        .enumerate()
+        .filter_map(|(block, chunks)| {
+            let block = block as u64;
+            // Already a valid block
+            if chunks.len() == 1 && chunks[0].is_block() {
+                None
+            // Already a pseudo-block, we can keep the chunk as it is
+            } else if chunks.len() == 1 && chunks[0].is_pseudo_block() {
+                let new_chunk = chunks[0].clone();
+                let to_remove = HashSet::new();
+                let write_back = false;
+                Some((block, chunks.to_vec(), new_chunk, write_back, to_remove))
+            // Reshape those chunks as a single block
+            } else {
+                // Start and stop should be 0 and blocksize respectively
+                let start = chunks[0].start;
+                let stop = chunks.last().expect("A block cannot be empty").stop;
+                let new_chunk = Chunk::new(start, stop);
+                let to_remove = chunks.iter().map(|x| x.id).collect();
+                let write_back = true;
+                Some((block, chunks.to_vec(), new_chunk, write_back, to_remove))
+            }
+        })
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::file_operations::{prepare_read, prepare_reshape, prepare_resize, prepare_write};
+    use parsec_api_types::{Blocksize, ChunkID, DateTime, DeviceID, EntryID};
+    use parsec_client_types::{Chunk, LocalFileManifest};
+    use std::{collections::HashMap, str::FromStr};
+
+    fn padded_data(data: &[u8], start: i64, stop: i64) -> Vec<u8> {
+        assert!(start <= stop);
+        if start <= stop && stop <= 0 {
+            vec![0; (stop - start) as usize]
+        } else if 0 <= start && start <= stop {
+            data[start as usize..stop as usize].to_vec()
+        } else {
+            let mut result: Vec<u8> = vec![0; (-start) as usize];
+            result.extend_from_slice(&data[0..stop as usize]);
+            result
+        }
+    }
+
+    #[derive(Default)]
+    struct Storage(HashMap<ChunkID, Vec<u8>>);
+
+    impl Storage {
+        fn read_chunk_data(&self, chunk_id: ChunkID) -> &[u8] {
+            self.0.get(&chunk_id).unwrap()
+        }
+
+        fn write_chunk_data(&mut self, chunk_id: ChunkID, data: Vec<u8>) {
+            let result = self.0.insert(chunk_id, data);
+            assert!(result.is_none());
+        }
+
+        fn clear_chunk_data(&mut self, chunk_id: ChunkID) {
+            self.0.remove(&chunk_id);
+        }
+
+        fn read_chunk<'a>(&'a self, chunk: &Chunk) -> &'a [u8] {
```

Comment:
Useless lifetime

Reply:
don't clippy should find those useless lifetime ? https://rust-lang.github.io/rust-clippy/master/#needless_lifetimes